### PR TITLE
esp32/machine_i2s: Integrate new I2S IDF driver.

### PR DESCRIPTION
**Pull request by miketeachman**
The legacy I2S "shim" is removed and replaced by the new I2S driver. The new driver fixes a bug where mono audio plays only in one channel.
Application code size is reduced by 2672 bytes with this change. Tested on ESP32, ESP32+spiram, ESP32-S3 using example code from https://github.com/miketeachman/micropython-i2s-examples

### CODE REVIEW

**Comment by ricksorensen** [comment]
Thanks - 

Just tried this with SEEED XIAO ESP32-C3 - and it works.  I ran your easy_wav_player using your mono and stereo wav files  pulled from the local memory (so only 16-bit versions).  I used pcm5102 for I2S to audio out.

**Comment by miketeachman** [comment]
@ricksorensen Fantastic !  Thank you for testing this PR.

File: extmod/machine_i2s.c [hunk 1]
```diff
@@ -600,6 +600,7 @@ STATIC mp_uint_t machine_i2s_stream_write(mp_obj_t self_in, const void *buf_in,
 
         return size;
     } else { // blocking or asyncio mode
+
```
Comment (dpgeorge) [inline extmod/machine_i2s.c:603]: I think this blank line can be removed, to keep coding style consistent (eg with the "if" part of this statement).

File: extmod/machine_i2s.c [hunk 2]
```diff
@@ -623,6 +625,7 @@ STATIC mp_uint_t machine_i2s_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr_
 
         if (flags & MP_STREAM_POLL_RD) {
             if (self->mode != MICROPY_PY_MACHINE_I2S_CONSTANT_RX) {
+
```
Comment (dpgeorge) [inline extmod/machine_i2s.c:628]: This blank line can go.

File: extmod/machine_i2s.c [hunk 3]
```diff
@@ -632,21 +635,14 @@ STATIC mp_uint_t machine_i2s_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr_
...
-                    ret |= MP_STREAM_POLL_RD;
-                }
+            if (self->dma_buffer_status == DMA_MEMORY_NOT_EMPTY) {
+                ret |= MP_STREAM_POLL_RD;
             }
             #endif
         }
 
         if (flags & MP_STREAM_POLL_WR) {
+
```
Comment (dpgeorge) [inline extmod/machine_i2s.c:645]: Remove blank line.

File: extmod/machine_i2s.c [hunk 4]
```diff
@@ -657,20 +653,14 @@ STATIC mp_uint_t machine_i2s_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr_
...
-                    ret |= MP_STREAM_POLL_WR;
-                }
+
+            if (self->dma_buffer_status == DMA_MEMORY_NOT_FULL) {
+                ret |= MP_STREAM_POLL_WR;
             }
             #endif
         }
     } else {
+
```
Comment (dpgeorge) [inline extmod/machine_i2s.c:663]: Remove blank line.

File: extmod/machine_i2s.c [hunk 5]
```diff
@@ -657,20 +653,14 @@ STATIC mp_uint_t machine_i2s_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr_
...
-            // (which is an indication that at least one DMA buffer is available to be written)
-            // note:  timeout = 0 so the call is non-blocking
-            i2s_event_t i2s_event;
-            if (xQueueReceive(self->i2s_event_queue, &i2s_event, 0)) {
-                if (i2s_event.type == I2S_EVENT_TX_DONE) {
-                    // getting here means that at least one DMA buffer is now empty
-                    // indicating that audio samples can be written to the I2S object
-                    ret |= MP_STREAM_POLL_WR;
-                }
+
```
Comment (dpgeorge) [inline extmod/machine_i2s.c:656]: Remove blank line.

File: ports/esp32/machine_i2s.c [hunk 6]
```diff
@@ -306,23 +276,52 @@ STATIC void task_for_non_blocking_mode(void *self_in) {
...
+};
+
+i2s_event_callbacks_t i2s_callbacks_null = {
+    .on_recv = NULL,
+    .on_recv_q_ovf = NULL,
+    .on_sent = NULL,
+    .on_send_q_ovf = NULL,
+};
+
+
```
Comment (dpgeorge) [inline ports/esp32/machine_i2s.c:307]: only one blank line needed here (for consistency with the rest of this file)

File: ports/esp32/machine_i2s.c [hunk 7]
```diff
@@ -417,21 +419,25 @@ STATIC machine_i2s_obj_t *mp_machine_i2s_make_new_instance(mp_int_t i2s_id) {
...
+            vQueueDelete(self->non_blocking_mode_queue);
+            self->non_blocking_mode_queue = NULL;
+        }
+        self->is_deinit = true;
     }
-
-    self->i2s_event_queue = NULL;
 }
 
+
```
Comment (dpgeorge) [inline ports/esp32/machine_i2s.c:440]: blank line not needed

**Comment by dpgeorge** [comment]
Thanks for this, it looks good!

I think the line `CONFIG_I2S_SUPPRESS_DEPRECATE_WARN` can now be removed from `esp32/boards/sdkconfig.base`.

**Comment by miketeachman** [comment]
@dpgeorge  thanks for reviewing this PR.  I removed all the unnecessary blank lines.  The macro could be removed after I included the header file for the new driver.

**Comment by projectgus** [comment]
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.

**Comment by dpgeorge** [comment]
I'm currently rebasing this on latest master and about to merge it.

**Pull request merged**